spike_main/spike.cc

```diff
@@ -41,6 +41,7 @@ static void help(int exit_code = 1)
   fprintf(stderr, "  --debug-cmd=<name>    Read commands from file (use with -d)\n");
   fprintf(stderr, "  --isa=<name>          RISC-V ISA string [default %s]\n", DEFAULT_ISA);
   fprintf(stderr, "  --pmpregions=<n>      Number of PMP regions [default 16]\n");
+  fprintf(stderr, "  --pmpgranularity=<n>  PMP Granularity [default 0]\n");
```

Comment:
What are the units here? Bytes?

0 does not make sense as a granularity. In RISC-V, 4 bytes is the minimum PMP granularity.

Reply:
Spike already implements PMPs as per the riscv priv spec. [This](https://github.com/riscv-software-src/riscv-isa-sim/blob/5a1145742e701597eb45825855311dfad21232a6/riscv/processor.cc#L66) line sets the sets the default granularity. 4 is the minimum PMP granularity, and that is taken care of, by the macro `PMP_SHIFT`. In order to set a coarser grain, we would have to increment the value of that shift on the LSB side of pmpaddr registers. This CLI arg should contain the value to be added to `PMP_SHIFT` while setting the grain.

this line makes that change:
```
set_pmp_granularity(1 << (PMP_SHIFT+cfg->pmpgranularity));
```

If the granularity is to remain `4B`, then `0` should be present in the value `cfg->pmpgranularity`.
Please correct me if im wrong.

Reply:
Maybe the help message should have been:
```
fprintf(stderr, "  --pmpgranularity=<n>  PMP Granularity [default 0 (4B)]\n");
```
instead of:
```
fprintf(stderr, "  --pmpgranularity=<n>  PMP Granularity [default 0]\n");
```
Please let me know if this is correct, or if the whole method of setting the granularity is wrong in this PR.

Reply:
You are wrong because you are conflating **G** with PMP granularity. They are not synonymous. The PMP granularity is 2**(G+2) bytes which has a minimum of 4, not 0.

Several options here:
* Rename the command line argument to "PMP G" (which is not very self-explanatory except to people who have the PMP spec fresh in their minds)
* Calculate G from the byte value given on the command line
* Directly pass the byte value given on the command line to `set_pmp_granularity()`, which already has legality checks

The last option seems cleanest and easiest.

Reply:
You're right. Directly accepting the granularity in bytes seems like a much more straightforward approach. Thanks for pointing it out. I've made the changes like you suggested.